File: Robot/Rectangle.swift

struct Size: ObjectConvertible {
	var width: Double
	var height: Double
	
	init(_ width: Double, _ height: Double) {
		self.width = width;
		self.height = height;
	}
	
	// --- ObjectConvertible
	
	static func convertFromObject(_ object: Object) throws -> Size {
		var size = Size(0,0)
		try size.convertFromObject(object)
		return size
	}
	
	mutating func convertFromObject(_ object: Object) throws {
		var v = Vector2D(0,0);
		try v.convertFromObject(object);
		width = v.x;
		height = v.y;
	}
	
	func convertToObject() -> Object {
		return Vector2D(width, height).convertToObject();
	}
}

struct Rectangle: ObjectConvertible {
	var origin: Vector2D
	var size: Size
	
	init(_ x: Double, _ y: Double, _ width: Double, _ height: Double) {
		origin = Vector2D(x, y);
		size = Size(width, height);
	}
	
	init(origin: Vector2D, size: Size) {
		self.origin = origin
		self.size = size
	}
	
	mutating func move(_ dv: Vector2D) {
		origin.addInPlace(dv)
	}
	func moved(_ dv: Vector2D) -> Rectangle {
		return Rectangle(origin: origin.add(dv), size: size)
	}
	
	// MARK: ObjectConvertible
	
	static func convertFromObject(_ object: Object) throws -> Rectangle {
		var rect = Rectangle(0,0,0,0)
		try rect.convertFromObject(object)
		return rect
	}
	
	mutating func convertFromObject(_ object: Object) throws {
		guard let list = Helper.castToArrayOfObjects(object) else {
			// !!! can throw
			throw IllegalArgumentException("list is null");
		}
		
		if (list.count != 2) {
			throw IllegalArgumentException("list has not size == 2");
		}
		
		try self.origin.convertFromObject(list.get(0));
		try self.size.convertFromObject(list.get(1));
	}
	
	func convertToObject() -> Object {
		return .array([
			origin.convertToObject(),
			size.convertToObject()
			])
	}
	
	
}

File: Robot/Java/Rectangle.java
package Robot.Java;

import java.util.ArrayList;
import Robot.Java.Helper.ObjectConvertible;
import Robot.Java.Helper.Helper;

class Size implements ObjectConvertible {
	double width;
	double height;
	
	Size(double width, double height) {
		this.width = width;
		this.height = height;
	}
	
	Size copy() {
		return new Size(width, height);
	}
	
	// --- ObjectConvertible
	
	public void convertFromObject(Object object) throws IllegalArgumentException {
		Vector v = new Vector(0,0);
		// !!! can throw
		v.convertFromObject(object);
		width = v.x;
		height = v.y;
	}
	
	public Object convertToObject() {
		return new Vector(width, height).convertToObject();
	}
}

class Rectangle implements DrawableObject {
	Vector origin;
	Size size;
	
	Rectangle(Vector origin, Size size) {
		this.origin = origin;
		this.size = size;
	}
	
	Rectangle(double x, double y, double width, double height) {
		origin = new Vector(x, y);
		size = new Size(width, height);
	}
	
	Rectangle moved(Vector dv) {
		return new Rectangle(origin.add(dv), size.copy());
	}
	
	// --- ObjectConvertible
	
	public void convertFromObject(Object object) throws IllegalArgumentException {
		ArrayList<Object> list = Helper.castToArrayListOfObjects(object);
		// !!! can throw
		Helper.throwIllegalErgumentExceptionIfNull("list is null", list);
		
		if (list.size() != 2) {
			throw new IllegalArgumentException("list has not size == 2");
		}
		
		// !!! can throw
		this.origin.convertFromObject(list.get(0));
		// !!! can throw
		this.size.convertFromObject(list.get(1));
	}
	
	public Object convertToObject() {
		ArrayList<Object> list = new ArrayList<>();
		list.add(origin.convertToObject());
		list.add(size.convertToObject());
		return list;
	}
	
	// - Helper
	
	public static Rectangle fromObject(Object obj) throws IllegalArgumentException {
		Rectangle rect = new Rectangle(0,0,0,0);
		// !!! can throw
		rect.convertFromObject(obj);
		return rect;
	}
	
	// --- Drawable
	
	public MyColor getColor() { return null; }
	public Line getLine() { return null; }
	public Rectangle getRectangle() { return this; }
	public Polygon getPolygon() { return null; }
	public MyImage getImage() { return null; }
	
}
